diff --git a/src/hotspot/cpu/aarch64/globals_aarch64.hpp b/src/hotspot/cpu/aarch64/globals_aarch64.hpp
index f26ea2a8b..d8b198a0b 100644
--- a/src/hotspot/cpu/aarch64/globals_aarch64.hpp
+++ b/src/hotspot/cpu/aarch64/globals_aarch64.hpp
@@ -80,9 +80,7 @@ define_pd_global(intx, InitArrayShortSize, BytesPerLong);
 
 define_pd_global(bool, ThreadLocalHandshakes, true);
 
-#if defined(COMPILER1) || defined(COMPILER2)
 define_pd_global(intx, InlineSmallCode,          1000);
-#endif
 
 #define ARCH_FLAGS(develop, \
                    product, \
