From ae703c0e86c278359f1bddcf35ccba87b556d901 Mon Sep 17 00:00:00 2001
Date: Thu, 18 Mar 2021 12:37:11 +0000
Subject: [PATCH 4/4] Fix the memcpy symbol issue during JDK11 x64 build

---
 make/lib/Awt2dLibraries.gmk | 1 +
 1 file changed, 1 insertion(+)

diff --git a/make/lib/Awt2dLibraries.gmk b/make/lib/Awt2dLibraries.gmk
index 207a459ae..7b0441507 100644
--- a/make/lib/Awt2dLibraries.gmk
+++ b/make/lib/Awt2dLibraries.gmk
@@ -597,6 +597,7 @@ else
   $(eval $(call SetupJdkLibrary, BUILD_LIBHARFBUZZ, \
       NAME := harfbuzz, \
       EXCLUDE_FILES := $(LIBHARFBUZZ_EXCLUDE_FILES), \
+      EXTRA_FILES := $(LIBMEMCPY_FILES), \
       TOOLCHAIN := TOOLCHAIN_LINK_CXX, \
       CFLAGS := $(CFLAGS_JDKLIB) $(LIBHARFBUZZ_CFLAGS), \
       CXXFLAGS := $(CXXFLAGS_JDKLIB) $(LIBHARFBUZZ_CFLAGS), \
-- 
2.19.0

